src/worker.rs

```diff
@@ -60,74 +59,75 @@ impl<'a> SidekiqWorker<'a> {
         info!("worker '{}' start working", self.with_server_id(&self.id));
-        // main loop is here
-        let rx = self.rx.clone();
-        let clock = tick(Duration::from_secs(1));
+        let rx = self.rx.take().unwrap();
+        let mut clock = time::interval(Duration::from_secs(1));
         loop {
             select! {
-                default => {
-                    debug!("{} run queue once", self.id);
-                    match self.run_queue_once() {
-                        Ok(true) => self.processed += 1,
-                        Ok(false) => {}
-                        Err(e) => {
-                            self.failed += 1;
-                            warn!("uncaught error '{}'", e);
+                biased;
+                Ok(op) = rx.recv() => {
+                    match op {
+                        Operation::Terminate => {
+                            info!("{} Terminate signal received, exiting...", self.id);
+                            self.tx.send(Signal::Terminated(self.id.clone())).await.ok();
+                            debug!("{} Terminate signal sent", self.id);
+                            return;
                         }
-                    };
+                    }
                 },
-                recv(clock) -> _ => {
-                    // synchronize state
+                _ = clock.tick() => {
                     debug!("{} syncing state", self.id);
-                    self.sync_state();
+                    self.sync_state().await;
                     debug!("{} syncing state done", self.id);
                 },
-                recv(rx) -> op => {
-                    if let Ok(Operation::Terminate) = op {
-                        info!("{} Terminate signal received, exiting...", self.id);
-                        self.tx.send(Signal::Terminated(self.id.clone())).ok();
-                        debug!("{} Terminate signal sent", self.id);
-                        return;
-                    } else {
-                        unimplemented!()
-                    }
-                },
+            }
+            match self.run_queue_once().await {
+                Ok(true) => self.processed += 1,
+                Ok(false) => {}
+                Err(e) => {
+                    self.failed += 1;
+                    warn!("uncaught error '{}'", e);
+                }
             }
         }
     }
 
-
-    fn run_queue_once(&mut self) -> Result<bool> {
+    // Note: this blocks for the duration passed to `brpop`
+    async fn run_queue_once(&mut self) -> Result<bool> {
         let queues: Vec<String> = self.queues.iter().map(|q| self.queue_name(q)).collect();
 
-        let result: Option<(String, String)> = self.pool.get()?.brpop(queues, 10)?;
+        let mut result: Result<Option<(String, String)>> = Ok(None);
+        task::block_in_place(|| {
+            Handle::current().block_on(async {
+                result = self.redis.brpop(queues, 10).await.map_err(From::from);
```

Comment:
As far as I can tell, this is the correct way to run blocking operations such that it doesn't block other async tasks from running. Note, we intentionally want to preserve this blocking `brpop` because it allows workers to start working on new jobs as soon as they're available.

Reply:
Also note that the server, and each worker individually has its own Redis connection. If they shared the same connection this wouldn't work properly.